testutils/testutils.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2018 The Prometheus Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testutils
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+// GatherAndCompare retrieves all metrics exposed by a collector and compares it
+// to an expected output in the Prometheus text exposition format.
+// metricNames allows only comparing the given metrics. All are compared if it's nil.
+func GatherAndCompare(c prometheus.Collector, expected string, metricNames ...string) error {
```

Comment:
The interface and implementation couple two independent concerns: verifying the given collector and formatting a custom input format. Instead of this coupling and assuming all users will format their expectation in this special way, using the `io.Reader` interface would leave it to the user how to provide the expectation. For example, this allows projects with extensive end to end tests to simply provide a File reader to such fixtures.

The current custom sanitizer implementation in `removeUnusedWhitespace` can be provided as another public function which will satisfy the reader interface.

Reply:
It certainly makes sense to use an `io.Reader` instead of a `string` as input.

I don't quite follow the line of argument with the `removeUnusedWhitespace`, though. In fact, now that I have looked more closely, I got the suspicion the `removeUnusedWhitespace` is not even needed. The provided text for "expected" is parsed with the text format parses. (Thus, the parser should eat all the spurious whitespace.) For the diff output, we then re-create the text format from the protobuf representation and get a sanitized output anyway.

Reply:
Even better if we don't need it.

Reply:
I played with the tests, and it really looks to me we can just remove `removeUnusedWhitespace`. :o)